workflows/cloudify_system_workflows/snapshots/credentials.py

```diff
@@ -69,121 +70,136 @@ def _get_hosts():
-                                   agent_key_path_dict):
-        restored_agent_key_path = \
-            self._restore_agent_key_from_dump(dump_cred_dir,
-                                              restored_cred_dir,
-                                              dep_node_id)
-        db_agent_key_path = agent_key_path_dict[dep_node_id]
-
-        if os.path.isfile(db_agent_key_path):
-            self._handle_existing_agent_key_path(
-                restored_agent_key_path,
-                db_agent_key_path
-            )
-        else:
-            copy_file(restored_agent_key_path, db_agent_key_path)
-
-    @staticmethod
-    def _handle_existing_agent_key_path(restored_key_path,
-                                        db_key_path):
-        """Handle the case where the agent key already exists; either ignore
-        it if the contents are identical, or raise an error if they aren't
-
-        :param restored_key_path: The key path as restored from the dump
-        :param db_key_path: The key path as retrieved from the DB
-        """
-        with open(db_key_path) as key_file:
-                content_1 = key_file.read()
-        with open(restored_key_path) as key_file:
-            content_2 = key_file.read()
-        if content_1 != content_2:
-            raise NonRecoverableError(
-                'Agent key path already taken: {0}'.format(db_key_path)
-            )
-        ctx.logger.debug('Agent key path already exist: '
-                         '{0}'.format(db_key_path))
+        try:
+            shutil.copy(source, destination)
+        except IOError as e:
+            if e.filename == source:
+                ctx.logger.debug(
+                    "Key doesn't appear to be a file path. Skipping")
+            else:
+                raise
+
+
+def get_agent_config(node_properties):
+    """cloudify_agent is deprecated, but still might be used in older
+    systems, so we try to gather the agent config from both sources
+    """
+    cloudify_agent = node_properties.get('cloudify_agent', {})
+    agent_config = node_properties.get('agent_config', {})
+    agent_config.update(cloudify_agent)
+    return agent_config
+
+
+def candidate_key_names(path):
+    filtered = SECRET_STORE_AGENT_KEY_PREFIX + ''.join(
+        char if char in ALLOWED_KEY_CHARS else '_'
+        for char in path
+        )
+    yield filtered
+    for suffix in itertools.count(1):
+        yield '{name}_{suffix}'.format(name=filtered, suffix=suffix)
+
+
+def restore(tempdir, postgres):
+    dump_cred_dir = os.path.join(tempdir, CRED_DIR)
+    if not os.path.isdir(dump_cred_dir):
+        ctx.logger.info('Missing credentials dir: '
+                        '{0}'.format(dump_cred_dir))
+        return
+
+    client = get_rest_client()
+
+    tenants = [t.name for t in client.tenants.list()]
+
+    credential_dirs = set(os.listdir(dump_cred_dir))
+
+    for tenant in tenants:
+        client._client._set_header('Tenant', tenant)
```

Comment:
Yep, or perhaps it's better to switch back to constructing a client per tenant, rather than messing with underscored vars (there's no way to pass a tenant in per operation)?